src/main/java/org/testng/internal/XmlMethodSelector.java

```diff
@@ -56,8 +73,9 @@ public boolean includeMethod(
       init(context);
     }
 
-    if (m_expression != null) {
-      return m_bsh.includeMethodFromExpression(m_expression, tm);
+    if (script != null) {
+      // TODO could be improved if cache by language and/or expression
+      return getScriptSelector(script.getLanguage()).includeMethodFromExpression(script.getExpression(), tm);
```

Comment:
True. I updated the pull request but it is not as good as expected because the api only allow to eval the script each time.
I expected that it was possible to pre-eval the script (engine optimization?) and then just provide the updated context.